### ISSUE TITLE
grad method needs data in float format

### ISSUE THREAD

**Issue by mariajmolina** [ai2es/sharpness#9]
Hello @stockeh et al.! Looks like the [grad method ](https://github.com/ai2es/sharpness/blob/main/src/metrics.py#L20)needs the input data in float format, not integer format, otherwise it outputs the really tiny values (e.g., e-19). I am not sure what would make the most sense to address this -- maybe adding an assert call to check that the arrays are in float format in the function? 

![Screenshot 2023-09-18 at 2 48 35 PM](https://github.com/ai2es/sharpness/assets/24659645/73832a66-0d49-4f30-a42e-73fe4d31eaa9)

**Comment by zyjux** [comment]
I was able to get the grad method to return the same value for the same data in int and float dtypes, but only if I enforced the dtype and scaled appropriately. The initial data is int64, so skimage assumes the data lie in the range (-2^63, 2^63) and scales its answer appropriately, hence why it returns values on the order of 10^-18. If we convert to uint8, it correctly scales for the range (0, 255) and gives a reasonable answer. If we convert the initial data to np.float32 instead, it gives a much larger answer, since it is assuming that the data are in the range (0, 1), or (-1, 1). If I divide the values by 255 to get the result in the appropriate range, then we get the same (reasonable) answer from both float and int data.

![skimage_dtype_issue_ex](https://github.com/ai2es/sharpness/assets/68205917/02b944cb-e93b-4950-b930-4fa2aa78ca28)

I propose we make our dataloader.py file ensure that the dtype it outputs is appropriate (likely uint8) so that at least our built-in functions aren't causing this issue, and have either benchmark.py or __init__.py's methods run a check and print out warnings if the data are: 1) an int type with a max value that is far less than max supported by the dtype or 2) a float type with values outside (-1, 1).

**Comment by mariajmolina** [comment]
Thanks, @zyjux! My concern is that sklearn states that we should not apply `astype` in their explanation[ here.](URL)

> Functions in skimage are designed so that they accept any of these dtypes, but, for efficiency, may return an image of a different dtype (see [Output types](https://scikit-image.org/docs/stable/user_guide/data_types.html#output-types)). If you need a particular dtype, skimage provides utility functions that convert dtypes and properly rescale image intensities (see [Input types](https://scikit-image.org/docs/stable/user_guide/data_types.html#input-types)). You should never use astype on an image, because it violates these assumptions about the dtype range:

They provide several utility functions that I am testing now and will reply shortly.

**Comment by mariajmolina** [comment]
I can get skimage to trigger a warning when I do the following, which I think would be the preferred behavior to warn users?

![Screenshot 2023-10-30 at 8 36 29 AM](https://github.com/ai2es/sharpness/assets/24659645/41d390c4-3d57-439e-87a5-edb8b51fda22)

Then, to rescale, skimage provides the following utility.

![Screenshot 2023-10-30 at 8 38 01 AM](https://github.com/ai2es/sharpness/assets/24659645/2aff0145-396a-4785-9f38-c1efee78e99b)

However, as they note in skimage, we may not want to rescale the float values just to fit the float range, because that may result in undesired behavior. E.g., maybe there is very light rain, but rescaling to 0 and 1 will create artificial sharp edges in an image.

> However, in some cases, the image values represent physical measurements, such as temperature or rainfall values, that the user does not want rescaled.

**Comment by zyjux** [comment]
At this point, I'm a bit tempted to simply drop skimage for this grad operator and implement it using OpenCV's Sobel function, as used in @jsschreck's gradients.py script. Regardless, we should probably standardize on one package implementation of the Sobel operator, either OpenCV or skimage, and if skimage is having all these issues maybe we should just not use it.

It seems like the issue is that skimage is making some sort of internal assumption about how to scale its output based on the input data type. So, to get consistent results, we'd need to have some sort of standardization on data type or rescaling based on data type, but rescaling gets into huge issues. In particular, if we scale each image separately, we're losing a lot of intensity information, and if we determine scale parameters based on some prior set of data then we may run into issues where future data violates the range assumptions we're trying to scale to. Now, it does seem like the skimage Sobel function will give results for float data that falls outside of the (0, 1) range, so we could simply encourage users to convert their data to float types and scale to (0, 1) based on "training" data (as not everyone is going to use this for ML training tasks necessarily) and if future samples go outside that range somewhat, the method should still handle that sanely.

**Comment by mariajmolina** [comment]
Totally agree with your sentiment!

> At this point, I'm a bit tempted to simply drop skimage for this grad operator and implement it using OpenCV's Sobel function, as used in @jsschreck's gradients.py script. Regardless, we should probably standardize on one package implementation of the Sobel operator, either OpenCV or skimage, and if skimage is having all these issues maybe we should just not use it.

Makes sense to me to focus on using just one package -- plus that would reduce the number of dependencies.

> ... but rescaling gets into huge issues. In particular, if we scale each image separately, we're losing a lot of intensity information, and if we determine scale parameters based on some prior set of data then we may run into issues where future data violates the range assumptions we're trying to scale to. Now, it does seem like the skimage Sobel function will give results for float data that falls outside of the (0, 1) range, so we could simply encourage users to convert their data to float types and scale to (0, 1) based on "training" data (as not everyone is going to use this for ML training tasks necessarily) and if future samples go outside that range somewhat, the method should still handle that sanely.

This also seems sensible. So perhaps making the change to only OpenCV is not urgent, but something to keep in mind and prioritize eventually. Thanks again for your help with sorting this issue out!

**Comment by zyjux** [comment]
As of commit 2be5b8adf6ad7246026374255935d85689d2730a, we have standardized on float data type by removing the skimage sobel function, as it was a duplicate of the already-implemented mean gradient magnitude metric which used opencv. At this point, input data should be floats; if they are not already floats, the package internally converts the inputs to float data for the routines that require it.
